### ISSUE TITLE
Restructure saving process as a pipeline

### ISSUE THREAD

**Issue by jacobsa** [jacobsa/comeback#21]
**Warning to self**: Do not start on this until after fixing up the mystery truncated blobs (#18) and successfully verifying the corpus (#14). Otherwise you're asking for pain.

I spent some time today trying to understand why the hacked up program that maps SHA1 to MD5 and CRC32 (at b55e0856c4d4210a8abd2495bb446021014d4bf1) is so damn slow.

Similar question: why is a normal `comeback save` when the caches are gone so slow? Ideally we wouldn't need caches at all, although the bandwidth of a hard drive might mean we still do.

The issue in this case appears to at least in part be due to encrypting on only a single core. We're forced into this by the blob store API, which is synchronously used by the file and directory saves, and synchronously returns a score.

Rather than making the file and directory save APIs more complicated and still coming up with a program that is hard to be confident doesn't have surprise bottlenecks, let's try explicitly modeling this as a pipeline of concurrent actors where each stage is sized to its relevant bottleneck:

0. Set `GOMAXPROCS` to N, where N is close to  `runtime.NumCPU()`.

1.  Do the following concurrently, communicating across stages with channels:
    * On a single goroutine, stat everything in the base directory for the backup with a BFS or DFS.
    
      We use only one goroutine here because this is not trivial to parallelize and the bottleneck here by far should be the hard drive, which has only one spindle.)
      
    * On a single goroutine, look up each stat'd file in our cache mapping from stat info to scores. Fill in scores for each hit.
    
      We use only one goroutine here because access to the cache must be synchronized. In fact, we should save CPU here by requiring external synchronization (we need none, because we are on one goroutine).

    * On a single goroutine, accumulate all into some in memory structure. This might be indexed by path depth (see below).

2.  For each path depth D, starting with the max and going downward:
    * On a single goroutine, read the contents of each file at depth D without scores already present, chunk by chunk. Pass on records that include the file path, the chunk, and the chunk index.
    
      We use only one goroutine because the bottleneck here is the hard drive, which has only one spindle.

    * On a single goroutine, assemble a listing for each directory at depth D and serialize it. This should be possible because D+1 is already done (TODO(jacobsa): think hard about this). Pass on records that include the directory path and serialization.
    
    * On N-1 goroutines, accept file/directory chunks and encrypt them.
    
    * On a single goroutine:
      *  Accept these and see whether they are already in GCS or in-flight at a later stage, using a set of existing scores. If so, short circuit.
      *  Accept output from the GCS goroutines below, updating the set of scores.
      *  Periodically write out a state file?
      *  This goroutine is the eventual sink, so it will probably also be responsible for synchronizing on the completion of everything.

    * On N-1 goroutines, accept encrypted data and compute SHA-1, MD5, and CRC32C.
    
    * On a GCS BDP-tuned number of goroutines, accept these records and create GCS objects from them. Write the result back to the goroutine checking for in-flight/pre-existing objects.

3. The score for the base directory is the outcome of the backup.

Some notes:

* Just throw all of this in a single package for now; restructure later.
* Use type names for plaintext vs. ciphertext and various checksums, for easy compiler checking.
* Write unit tests for each pipeline stage.
* Make sure to verify the output of this process:
  * First run on an already-saved backup and make sure the score is the same.
  * Then run on a brand new small corpus, verify, restore, and make sure everything is sane.

**Comment by jacobsa** [comment]
For what it's worth, the hacked up program that prints a SHA1 -> CRC32C/MD5 map at b55e0856c4d4210a8abd2495bb446021014d4bf1 takes about 5 hours to deal with 565,851 blobs. Hopefully we can go faster than that.

**Comment by jacobsa** [comment]
Pipeline for restore might look something like this:

* Do a BFS through the DAG starting with backup roots, remembering what has already been processed or is in-flight.
* Accumulate a slice containing all expected paths and their stat info. Hold back hard links for now, segregating them.
* Process all files, setting permissions, owners, etc.
* Process all hard links (which must be files, confirm this in code).
* Process all directories, setting permissions, owners, etc.

**Comment by jacobsa** [comment]
For both saving and restoring, we need to traverse the DAG defined by the file system tree (plus outgoing edges from files to other files for hard links when restoring), visiting successors before predecessors:

*   When saving, we need to ensure we have scores for files/directories before making and saving the listing for their parents.

*   When restoring, we'll want to create the directory tree with all `0700` directories, then fill in children, then "seal off" parents if their permissions need to change.

Algorithm for doing this in parallel, given the DAG structure:

*   Mark all nodes initially with the number of successors they have and the list of their predecessors.

*   Start workers that process nodes that are marked with zero. When they're done with a node, they should decrement each predecessor.

*   If there is no cycle, every node will eventually become zero.

Probably rename graph.Traverse to graph.Discover (since it doesn't have the structure up front and makes no order guarantees beyond discovery order). Call this one graph.TraverseDAG.

**Comment by jacobsa** [comment]
Luckily as of f147fc62639d34ae154449015d3e1d0be5b43d7b we don't support preserving hard links, so all of the complications related to those above can go away.

**Comment by jacobsa** [comment]
Attempt at a plan for the `graph` package:

* Define `type NodeInfo interface{}` to be any info the user needs to knoe about a node. Within a graph, all values must be comparable (so that they can be used in map keys).

* `type SuccessorFinder interface` accepts `context.Context, NodeInfo` and returns `[]NodeInfo, error`.

* `func ExploreGraph` takes `SuccessorFinder`, `roots []NodeInfo`, and a parallelism value. Writes all nodes accessible from `roots` to `chan<- *Node` Makes no guarantee about the order. The implementation is mostly like `Traverse` today, including keeping lots of state, and the latter should be deleted.

* `func ReverseTopsortTree` accepts `SuccessorFinder`, `root NodeInfo`. Writes all notes in the tree rooted at `root` to `chan<- *Node`, with children guaranteed to be before parents. No parallelism.

*   `func VisitDAG` accepts `SuccessorFinder`, `<-chan NodeInfo`, `visitor func(context.Context, NodeInfo) error`, and parallelism. The channel must be topsorted: predecessors before successors. It also must not contain duplicates. It is legal for only a subgraph to be represented. Calls `visitor` once for each node, guaranteeing all predecessors have already been processed.

    Internally, uses the algorithm described [above](https://github.com/jacobsa/comeback/issues/21#issuecomment-103731287), except with the dependency order reversed. When a node comes through the channel, we increment a count for its successors. When a node finishes, we decrement all of the counts we incremented. We start nodes when their count is zero.

These will be used as follows:

*   For saving, `ReverseTopsortTree` will be used with a `SuccessorFinder` that lists the child names in the directory. This will give a channel containing paths with children before parents. That can be transformed into a channel of stat structs, where directory structs have back references to their children.
    
    `VisitDAG` can be applied to that channel with a `SuccessorFinder` of opposite polarity (basically just `path.Dir`) and a visitor that does everything needed for each one: read file if scores not cached, encrypt, write to GCS, etc. Each one of these steps can be guarded with a semaphore, and the overall parallelism for `VisitDAG` can be set to approximately the sum of the semaphore capacities.

*   Similarly for verifying, where today's `verify.visitor` has its behavior split in half: a `SuccessorFinder` that just reads and parses directories, and a visitor that doesn't anything additional necessary to verify.

**Comment by jacobsa** [comment]
Once again I've overcomplicated the graph package. Better:

* We only deal with DAGs.
* `DependencyResolver` interface lists dependencies for a node.
* `Visitor` accepts a node and returns an error.
* Single function `VisitDAG` takes one of each, and a list of known nodes. Visits all dependencies of those nodes and the nodes themselves. Dependencies before dependants.
* Guarantees that for each node there will be at most a single call to resolver and visitor for that node (i.e. no synchronization needed for writing to structs pointed to by nodes).
* Guarantees that dependency resolver is called at most once for each node, so that it need not be idempotent.
* Guarantees that visitor is called at least once for each node.
* (I think these two imply the concurrency guarantee?)
* Works by keeping track of outstanding count of dependencies for each node, processing those that hit zero.
* For verification: dependency resolver reads directory blobs. Visitor confirms score exists, and optionally reads file blobs.
* For saving: dependency resolver reads directories. Visitor consults score map, saves files if necessary, saves directories.
* For restoring: dependency resolver reads directory blobs. Visitor creates parent directories and writes directories, files, symlinks and sets permissions.

**Comment by jacobsa** [comment]
Still to do:

* Compare backup speeds and results on the same corpus between 2ead6ca and f1e9145. If there's a speed problem with cached scores, consider removing parallelization for the dependency resolver in `dag.Visit` (perhaps by taking a separate parallelism argument for that vs. visiting) to see if it helps us avoid throwing away disk locality.

* Try `comeback restore` on a smallish backup from the `issue-21` branch, then see what `rsync` says about it.

* Merge `issue-21` into `master`.

* Follow up on TODOs and delete all the dead code possible.
